### ISSUE TITLE
Cloud Backup: Nextcloud users with UUID

### ISSUE THREAD

**Issue by StillLoading** [opnsense/core#4080]
With the current configuration options it is not possible to use Nextcloud with LDAP/AD user integration in its default configuration.

By default the Nextcloud LDAP user backend uses the UUID attribute to add users. This is probably true for other backends too. As a consequence the webdav path does not include the username but the UUID e.g. https\://nextcloud.example.com/remote.php/dav/files/\<UUID>

**To Reproduce**
Nextcloud installation with LDAP backend (probably others that use UUIDs for created users).
1. Setup Nextcloud backup with LDAP user

**Relevant log files**

Throws http 404.

`config[12996]: {"url":"https:\/\/nextcloud.example.com\/remote.php\/dav\/files\/<username>\/","content_type":"application\/xml; charset=utf-8","http_code":404,"header_size":1102,"request_size":182,"filetime":-1,"ssl_verify_result":0,"redirect_count":0,"total_time":1.539546,"namelookup_time":0.347217,"connect_time":0.347459,"pretransfer_time":0.37096,"size_upload":0,"size_download":224,"speed_download":145,"speed_upload":0,"download_content_length":-1,"upload_content_length":-1,"starttransfer_time":1.539463,"redirect_time":0,"redirect_url":"","primary_ip":"<removed>","certinfo":[],"primary_port":443,"local_ip":"<removed>","local_port":60632}`

**Describe the solution you'd like**

Maybe instead of entering the base path we could opt to enter the full webdav path. This would also solve #3988 

**Environment**

OPNsense 20.1.5-amd64
FreeBSD 11.2-RELEASE-p18-HBSD
OpenSSL 1.1.1f 31 Mar 2020

**Comment by fabianfrz** [comment]
I do not own an LDAP server so I cannot test / reproduce that. Is there an documented way to query the uuid from nextcloud?

**Comment by StillLoading** [comment]
> I do not own an LDAP server so I cannot test / reproduce that. Is there an documented way to query the uuid from nextcloud?

Here's the explanation from the LDAP user backend.

> **Internal Username**
By default the internal username will be created from the UUID attribute. It makes sure that the username is unique and characters do not need to be converted. The internal username has the restriction that only these characters are allowed: [ a-zA-Z0-9_.@- ]. Other characters are replaced with their ASCII correspondence or simply omitted. On collisions a number will be added/increased. The internal username is used to identify a user internally. It is also the default name for the user home folder. It is also a part of remote URLs, for instance for all *DAV services. With this setting, the default behavior can be overridden. Leave it empty for default behavior. Changes will have effect only on newly mapped (added) LDAP users.

So basically the Nextcloud username is the UUID. I could confirm this by entering the UUID into the "User Name" configuration in opnsense. After that I got a HTTP 401, which is correct since the UUID is not allowed as a login attribute in my case.

**Comment by fabianfrz** [comment]
no, that means that it is derived from the UUID unless otherwise configured. So I need to login and query the real username using an API endpoint.

https://docs.nextcloud.com/server/12.0/admin_manual/configuration_user/user_auth_ldap.html

**Comment by StillLoading** [comment]
You mean something like this ? https://docs.nextcloud.com/server/15/developer_manual/client_apis/LoginFlow/index.html#obtaining-the-login-credentials

**GET 'https\://nextcloud.example.com/ocs/v1.php/cloud/user' -H "OCS-APIRequest: true"**

```xml
<?xml version="1.0"?>
<ocs>
 <meta>
---buch of stuff---
 </meta>
 <data>
---This looks prominising---
  <enabled>1</enabled>
  <storageLocation>/vol/nextcloud_data/cf67c556-9012-22e7-9da5-5e4ba3d622e1</storageLocation>
  <id>cf67c556-9012-22e7-9da5-5e4ba3d622e1</id>
...
more stuff
 </data>
</ocs>
```

**Comment by fabianfrz** [comment]
this looks useful, so I can use the id attribute to detect an external user.

what does the backend tag say?

**Comment by StillLoading** [comment]
> this looks useful, so I can use the id attribute to detect an external user.
> 
> what does the backend tag say?

Yes the id tag contains the internal username. 

This is the backend tag from my LDAP user.
```xml
<backend>LDAP</backend>
```
Normal users have **Database** as backend. I couldn't find a list of possible values but I expect there are a few more.

**Comment by StillLoading** [comment]
I just thought of a way you can test this without an LDAP server. By default Nextcloud users can login with their E-Mail. So instead of entering the username in opnsense enter the E-Mail and you would have basically the same situation where username != loginname and != webdav folder.

**Comment by fabianfrz** [comment]
@StillLoading I fixed in in the linked PR. please apply it and respond if it works. For me with the email address it worked with local database.

Kind regards

Fabian

**Comment by StillLoading** [comment]
@fabianfrz Just tested it and it's working! Thanks!
